### PR TITLE
`qml.ctrl` uses new `Controlled` op instead of `ControlledOperation`

### PR DESCRIPTION
**Pull request by albi3ro**
This PR accomplishes three main things:

* substitutes `Controlled` for `ControlledOperation` in `qml.ctrl`.
* Removes `ControlledOperation`
* Allows `qml.ctrl` to accept a single operator.

I initially wanted to make this multiple PR's, but `ControlledOperation` had no unittests.  Removing it from `ctrl` would leave it completely untested. I didn't want to write a bunch of unittests for something I was about to delete.

### CODE REVIEW

File: pennylane/ops/op_math/control.py [hunk 1]
```diff
@@ -17,284 +17,100 @@
...
 
-            op = qml.ctrl(ops, 2, control_values=0)
-            op(params=[0.1, 0.2])
+    @wraps(op)
+    def wrapper(*args, **kwargs):
+        tape = qml.transforms.make_tape(op)(*args, **kwargs)
 
-        This is equivalent to the following.
+        # flip control_values == 0 wires here, so we don't have to do it for each individual op.
+        flip_control_on_zero = (len(tape) > 1) and (control_values is not None)
```
[the remark this answers is not in the document]
Comment (AlbertMitjans) [reply pennylane/ops/op_math/control.py:96]: What if:
```python
def circuit():
    qml.PauliX(0)

ctrl_circuit = qml.ctrl(circuit, control=1, control_values=0)
```

File: pennylane/ops/op_math/controlled_class.py [hunk 2]
```diff
@@ -408,6 +410,16 @@ def __new__(cls, *_, **__):
...
 
+    # pylint: disable=too-many-function-args
+    def __init__(
+        self, base, control_wires, control_values=None, work_wires=None, do_queue=True, id=None
+    ):
+        super().__init__(base, control_wires, control_values, work_wires, do_queue, id)
+        # check the grad_recipe validity
+        if self.grad_recipe is None:
+            # Make sure grad_recipe is an iterable of correct length instead of None
+            self.grad_recipe = [None] * self.num_params
```
Comment (AlbertMitjans) [inline pennylane/ops/op_math/controlled_class.py:421]: What is this for?

File: pennylane/ops/op_math/control.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlbertMitjans) [reply pennylane/ops/op_math/control.py:96]: Also, I believe the drawing of a controlled operation with `control_values=0` is not correct:
```python
dev = qml.device("default.qubit", wires=2)

@qml.qnode(dev)
def circuit():
    qml.ctrl(qml.PauliX(0), control=1, control_values=0)
    return qml.probs(0)

print(qml.draw(circuit)())
```
```
0: ─╭X─┤  Probs
1: ─╰○─┤  
```
Shouldn't this be:
```
0: -──╭X───┤  Probs
1: ─X─╰○─X─┤   
```
On another topic, is there a way to draw an operator without having to build a qnode?

File: pennylane/ops/op_math/controlled_class.py [hunk 3]
```diff
@@ -171,6 +171,8 @@ def __init__(
                 )
                 control_values = [(x == "1") for x in control_values]
 
+            control_values = [control_values] if isinstance(control_values, int) else control_values
```
Comment (AlbertMitjans) [inline pennylane/ops/op_math/controlled_class.py:174]: Could we cast this into a `bool`?

File: tests/ops/functions/test_eigvals.py [hunk 4]
```diff
@@ -106,7 +106,7 @@ def test_ctrl(self):
         """Test that the ctrl is correctly taken into account"""
         res = qml.eigvals(qml.ctrl(qml.PauliX, 0))(wires=1)
         expected = np.linalg.eigvals(qml.matrix(qml.CNOT(wires=[0, 1])))
-        assert np.allclose(res, expected)
+        assert np.allclose(np.sort(res), np.sort(expected))
```
Comment (AlbertMitjans) [inline tests/ops/functions/test_eigvals.py:109]: Why do we obtain different ordering?

File: tests/ops/functions/test_is_commuting.py [hunk 5]
```diff
@@ -170,28 +170,11 @@ def test_cnot_cz(self, wires, res):
     def test_cz_mcz(self, wires, res):
         """Commutation between CZ and MCZ."""
 
-        def z():
-            qml.PauliZ(wires=wires[1][2])
+        op1 = qml.ctrl(qml.PauliZ(wires=wires[1][2]), control=wires[1][:-1])
```
Comment (AlbertMitjans) [inline tests/ops/functions/test_is_commuting.py:173]: Maybe use `wires[1][-1]` instead of `wires[1][2]` to make sure that `PauliZ` and `control` don't have overlapping wires.

File: tests/ops/functions/test_is_commuting.py [hunk 6]
```diff
@@ -204,32 +187,11 @@ def z():
...
-    @pytest.mark.parametrize(
-        "wires,res",
-        [
-            ([[0], [0, 1, 2]], True),
-            ([[1], [0, 1, 2]], True),
-            ([[2], [0, 1, 2]], True),
-        ],
-    )
-    def test_mcz_rx(self, wires, res):
-        """Commutation between MCZ and RZ"""
```
Comment (AlbertMitjans) [inline tests/ops/functions/test_is_commuting.py:224]: Why did you delete this test (`MCZ` and `RZ`)?

File: tests/ops/op_math/test_control.py [hunk 7]
```diff
@@ -35,6 +33,8 @@ def make_ops():
         # Execute controlled version.
         cmake_ops()
 
+    expanded_tape = tape.expand()
```
Comment (AlbertMitjans) [inline tests/ops/op_math/test_control.py:36]: Why do we need to expand the tape?

File: tests/transforms/test_condition.py [hunk 8]
```diff
@@ -284,21 +274,16 @@ def test_ctrl_operationss_with_cond(self, terminal_measurement):
         ops = tape.operations
         target_wire = qml.wires.Wires(2)
 
-        assert len(ops) == 2
+        assert len(ops) == 3
```
Comment (AlbertMitjans) [inline tests/transforms/test_condition.py:277]: Why did this change?

**Review by AlbertMitjans — COMMENTED** [review]
Nice work!

Almost all comments are questions 😆

File: pennylane/ops/op_math/controlled_class.py (hunk 2 above)
Comment (albi3ro) [reply pennylane/ops/op_math/controlled_class.py:421]: I forgot to add this earlier.  Without it, we can't use parameter-shift even though the `parameter_frequencies` are defined.

File: tests/ops/functions/test_eigvals.py (hunk 4 above)
Comment (albi3ro) [reply tests/ops/functions/test_eigvals.py:109]: `Controlled` operations' eigenvalues are ones followed by base operator eigenvalues.  The eigenvalues of `CNOT` are determined by `qml.math.linalg.eigvals` of the matrix.  The output of `np.linalg.eigvals` is not necessarily ordered.

So even though all the eigenvalues are the same, they were in different orders.

File: tests/ops/functions/test_is_commuting.py (hunk 6 above)
Comment (albi3ro) [reply tests/ops/functions/test_is_commuting.py:224]: I combined `test_rx_mcz` and `test_mcz_rx` together.  We now test both orders in the same test.

File: tests/ops/op_math/test_control.py (hunk 7 above)
Comment (albi3ro) [reply tests/ops/op_math/test_control.py:36]: Otherwise, we would have `Controlled(RX(0.123, 0), control_wires=1)`.  Upon expansion, that operation is converted into `CRX`.

Maybe we could more eagerly convert a Controlled RX into `CRX`, but that's just a potential task for later.

If you notice below, the old code did request `expanded = ctrl_op.expand()`.

File: tests/transforms/test_condition.py (hunk 8 above)
Comment (albi3ro) [reply tests/transforms/test_condition.py:277]: Before, the `ControlledOperation` wrapped a tape that had two operations inside it.  Now we have a `Controlled` for each of those two operations.

File: tests/ops/op_math/test_control.py (hunk 7 above)
Comment (AlbertMitjans) [reply tests/ops/op_math/test_control.py:36]: Is this conversion really an "expansion"? I feel like this should be in the grouping functionality.

On the other hand, will we remove the `CRX` and all other `C...` gates?

**Comment by Jaybsoni** [comment]
[sc-22447]

File: tests/ops/op_math/test_control.py (hunk 7 above)
Comment (albi3ro) [reply tests/ops/op_math/test_control.py:36]: "expansion" is kind of a catch-all for "convert to a form more likely to be supported by a device".  

I have been thinking about doing this type of conversion more eagerly instead of waiting for the decomposition to be requested, but that might be a later PR.

File: tests/ops/functions/test_is_commuting.py (hunk 5 above)
Comment (albi3ro) [reply tests/ops/functions/test_is_commuting.py]: done.

File: tests/ops/op_math/test_controlled_op.py [hunk 9]
```diff
@@ -846,3 +872,99 @@ def pow(self, z):
         assert pow_op.control_wires == self.control_wires
         assert pow_op.control_values == self.control_values
         assert pow_op.work_wires == self.work_wires
+
+
+@pytest.mark.parametrize("diff_method", ["backprop", "parameter-shift", "finite-diff"])
+class TestDifferentiation:
```
Comment (Jaybsoni) [inline tests/ops/op_math/test_controlled_op.py:878]: We could also test the differentiation of a controlled quantum function

**Review by Jaybsoni — COMMENTED** [review]
Looks great! Just some small comments!

File: doc/releases/changelog-dev.md [hunk 10]
```diff
@@ -153,6 +153,9 @@
 
 <h3>Improvements</h3>
 
+* `qml.ctrl` now uses `Controlled` instead of `ControlledOperation`.  The new `Controlled` class
+  wraps individual `Operator`'s instead of a tape.  It provides improved representations and integration.
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:157]:
```suggestion
  wraps individual `Operator`'s instead of a tape.  It provides improved representations and integration.
  [(#2990)](https://github.com/PennyLaneAI/pennylane/pull/2990)
```

File: doc/releases/changelog-dev.md [hunk 11]
```diff
@@ -281,6 +284,9 @@
   error. To definitively determine whether or not an operator is hermitian, use `qml.is_hermitian`.
   [(#2960)](https://github.com/PennyLaneAI/pennylane/pull/2960)
 
+* The `ControlledOperation` class is removed.  This was a developer-only class, so the change should not be evident to
+  any users. It is replaced by `Controlled`.
+
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply doc/releases/changelog-dev.md:289]:
```suggestion
* The `ControlledOperation` class is removed.  This was a developer-only class, so the change should not be evident to
  any users. It is replaced by `Controlled`.
  [(#2990)](https://github.com/PennyLaneAI/pennylane/pull/2990)

```

File: pennylane/ops/op_math/control.py [hunk 12]
```diff
@@ -17,284 +17,100 @@
...
-            control_values=new_values + self.control_values,
-        )
-
-
-def ctrl(fn, control, control_values=None):
-    """Create a method that applies a controlled version of the provided method.
-
-    Args:
-        fn (function): Any python function that applies pennylane operations.
+        op (function or :class:`~.operation.Operator`): A single operator or a function that applies pennylane operators.
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/ops/op_math/control.py:29]: "Quantum function" isn't an actual type? Other docstrings just tend to say `function` (or `callable`).

File: pennylane/ops/qubit/parametric_ops.py [hunk 13]
```diff
@@ -2097,8 +2097,7 @@ def simplify(self):
         if _can_replace(p1, 0):
             return qml.CRZ((p0 + p2) % (4 * np.pi), wires=wires)
         if _can_replace(p0, np.pi) and _can_replace(p1, np.pi / 2) and _can_replace(p2, 0):
-            hadamard = qml.Hadamard
-            return qml.ctrl(hadamard, control=self.control_wires)(wires=target_wires)
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/ops/qubit/parametric_ops.py:2101]: :)

File: tests/ops/functions/test_is_commuting.py [hunk 14]
```diff
@@ -850,30 +792,6 @@ def test_operation_2_not_supported(self):
         with pytest.raises(qml.QuantumFunctionError, match="Operation PauliRot not supported."):
             qml.is_commuting(qml.PauliX(wires=0), qml.PauliRot(1, "X", wires=0))
 
-    def test_operation_1_multiple_targets(self):
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply tests/ops/functions/test_is_commuting.py:853]: This is testing the use of `is_commuting` on `ControlledOperation`.  Since we no longer have `ControlledOperation`, we no longer need the test.

File: tests/ops/op_math/test_control.py [hunk 15]
```diff
@@ -74,59 +72,9 @@ def my_op(a, b, c):
         qml.CRY(-0.123, wires=[5, 3]),
         qml.CRX(-0.789, wires=[5, 2]),
     ]
-    for tape in [tape1.expand(depth=2), tape2.expand(depth=1)]:
-        assert_equal_operations(tape.circuit, expected)
-
-
-class TestAdjointOutsideQueuing:
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply tests/ops/op_math/test_control.py:81]: This sort of thing is tested in the unittests for `Controlled`.  

Think of this block as tests for `ControlledOperation` rather than tests for `qml.ctrl`.

File: tests/ops/op_math/test_controlled_op.py [hunk 16]
```diff
@@ -382,6 +383,31 @@ def test_diagonalizing_gates(self):
...
+        # all variations on default control_values
+        op4 = Controlled(base, (1, 2))
+        op5 = Controlled(base, (1, 2), [True, True])
+        op6 = Controlled(base, (1, 2), [1, 1])
+        assert op4.hash == op5.hash
+        assert op4.hash == op6.hash
+
+        # ignore work wires
+        op7 = Controlled(base, (1, 2), [0, 1], work_wires="aux")
+        assert op7.hash == op1.hash
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply tests/ops/op_math/test_controlled_op.py:409]: I ignored work wires as they don't actually impact the operation itself, potentially just its decompositions. But easy enough to consider them too.

File: tests/ops/op_math/test_controlled_op.py (hunk 9 above)
Comment (albi3ro) [reply tests/ops/op_math/test_controlled_op.py:878]: If I have time, I'll add some in.  Trying to just get the normal case working again now...

File: pennylane/ops/op_math/control.py [hunk 17]
```diff
@@ -17,284 +17,100 @@
...
+    tensor(0.36235775, requires_grad=True)
+    >>> qml.grad(circuit)(x)
+    -0.9320390859672264
 
-        @qml.qnode(dev)
-        def my_circuit2():
-            # ops1(params=[0.123, 0.456])
-            qml.CRX(0.123, wires=[1, 0])
-            qml.CRZ(0.456, wires=[1, 3])
+    :func:`~.ctrl` works on both functions, shown above, and individual :class:`~.operation.Operator`'s.
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/ops/op_math/control.py:63]: qml.RX is a callable and can thus be treated as a quantum function.

File: tests/ops/op_math/test_controlled_op.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (albi3ro) [reply tests/ops/op_math/test_controlled_op.py]: Now work wires influence the hash.

File: pennylane/ops/op_math/controlled_class.py [hunk 18]
```diff
@@ -416,6 +429,16 @@ def __new__(cls, *_, **__):
...
 
+    # pylint: disable=too-many-function-args
+    def __init__(
+        self, base, control_wires, control_values=None, work_wires=None, do_queue=True, id=None
+    ):
+        super().__init__(base, control_wires, control_values, work_wires, do_queue, id)
+        # check the grad_recipe validity
+        if self.grad_recipe is None:
+            # Make sure grad_recipe is an iterable of correct length instead of None
+            self.grad_recipe = [None] * self.num_params
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/ops/op_math/controlled_class.py:455]: Well, we already can't take a parameter shift derivative of Sum, Prod, or SProd, so I'm not too concerned.

File: pennylane/ops/op_math/controlled_class.py (hunk 3 above)
Comment (AlbertMitjans) [reply pennylane/ops/op_math/controlled_class.py:174]:
```suggestion
            control_values = [bool(control_values)] if isinstance(control_values, int) else [bool(control_value) for control_value in control_values]
```

Comment (AlbertMitjans) [reply pennylane/ops/op_math/controlled_class.py:174]: Also, what is `control_values` is a bool?

**Review by AlbertMitjans — COMMENTED** [review]
Just one small change:

**Review by AlbertMitjans — APPROVED** [review]
Approved! (sorry my suggestion broke the formatting check :S)

File: pennylane/ops/op_math/control.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/ops/op_math/control.py:29]: ahh, brain fart! yes we are stating the type! We don't need to include it.

**Review by Jaybsoni — APPROVED** [review]
Looks good to me 👍🏼!
